kate/recovery/src/com.rs

```diff
@@ -1,39 +1,115 @@
-use std::{convert::TryInto, ops::Range};
+use std::{collections::HashMap, convert::TryInto, ops::Range};
 
+use anyhow::{anyhow, Result};
 use dusk_bytes::Serializable;
 use dusk_plonk::{fft::EvaluationDomain, prelude::BlsScalar};
 
 // TODO: Constants are copy from kate crate, we should move them to common place
 pub const DATA_CHUNK_SIZE: usize = 31;
 const PADDING_TAIL_VALUE: u8 = 0x80;
 
-// Reconstructs app extrinsics from extrinsics layout and data
+pub struct MatrixDimensions {
+	pub rows: usize,
+	pub cols: usize,
+	pub chunk_size: usize,
+}
+
+fn map_cells(
+	dimensions: &MatrixDimensions,
+	cells: Vec<Cell>,
+) -> Result<HashMap<u16, HashMap<u16, Cell>>> {
+	let mut result: HashMap<u16, HashMap<u16, Cell>> = HashMap::new();
+	for cell in cells {
+		let row = cell.row;
+		let col = cell.col;
+		if row as usize > dimensions.rows * 2 || col as usize > dimensions.cols {
+			return Err(anyhow!("Invalid cell (col {}, row {})", col, row));
+		}
+		match result.get_mut(&col) {
+			None => {
+				let mut cells = HashMap::new();
+				cells.insert(row, cell);
+				result.insert(col, cells);
+			},
+			Some(cells) => {
+				if cells.contains_key(&cell.row) {
+					return Err(anyhow!("Duplicate cell found"));
+				}
+				cells.insert(cell.row, cell);
+			},
+		}
```

Comment:
`Entry` interface can simplify this.
```suggestion
	        let cells = result.entry(col).or_insert_with(HashMap::new);
		if cells.insert(cell.row, cell).is_some() {
			return Err(anyhow!("Duplicate cell found"));
		}
```
```

Reply:
Very nice 🥇